### DISCUSSION TITLE
Creation of NameSpace with external variables

### DISCUSSION THREAD

**Discussion by rafaone** [eclipse-milo/milo#1156] [General]
When I creating a name space, create a method to automate the creation of nodes,
```
 void addNodeX(UaObjectNode pNode, String name, Object value, EUInformation unit) {
			if(name==null){
				return;
			}else{
				if(name.equals("")){
					return;
				}
				
				 NodeId typeId = Identifiers.Integer;
				
				 if(value instanceof String){
					 typeId = Identifiers.String;
				 }else if (value instanceof Integer){
					 typeId = Identifiers.Integer;
				 }else if (value instanceof Double){
					 typeId = Identifiers.Double;
				 }else{
					 typeId = Identifiers.Integer;
				 }
				System.out.println("type id " + typeId.toString());
	            UaVariableNode node = new UaVariableNode.UaVariableNodeBuilder(getNodeContext())
	                .setNodeId(newNodeId( pNode.getNodeId().getIdentifier() + "." + name))
	                .setAccessLevel(AccessLevel.READ_ONLY)
	                .setBrowseName(newQualifiedName(name))
	                .setDisplayName(LocalizedText.english(name))
	                .setDataType(typeId)
	                .setTypeDefinition(Identifiers.BaseDataVariableType)
	                .setMinimumSamplingInterval(MIN_SAMPLING)
	                .build();
	            node.getFilterChain().addLast(
	                new AttributeLoggingFilter(),
	                AttributeFilters.getValue(
	                    ctx ->
	                        new DataValue(new Variant(value))
	                )
	            );
	            if(unit!=null){
	            	node.setEngineeringUnits(unit);
	            }
	            getNodeManager().addNode(node);
	            pNode.addComponent(node);
			}
	}
```

then I use this:
` addNodeX(wrd1OXI,"TEMP",(Main.thPrincipal.wrd1.opcUAInput.oxi.getTemp()),engUnitTemp);`

When I use this method, the node is created I can subscribe the node, but when I update the value, the value not reflect on client .

![image](https://github.com/eclipse/milo/assets/9700683/e1647c82-29f4-4be6-838f-1cb434d2839c)

the Temp keep on zero, but I've updated the variable reference on object.
`opcUAInput.tc.setTemp(2222);`

And the strange, if I insert the full add node on name space, it works, I just create method to avoid a lot of lines and keep maintenance better.

![image](https://github.com/eclipse/milo/assets/9700683/ad0ea636-4d32-4645-bcf6-56ce1e5ca5b0)

the red using addNodeX not update, if I do like a green block, it work.

I really dont know what is the problem, since is possible to install  the filter to watch a variable .

**Comment by kevinherron** [comment]
What does the implementation of `addNodeX` look like? From the signature it looks like you just pass the _value_ of that variable at the time you call it, not a reference to a function that can be called to get the current value. No mystery that you only get the initial value.

**Reply by rafaone** [reply]
ok, I need to use a muttable objects, can you delete this post?
